cmd/metal-api/internal/grpc/waitServer.go

```diff
@@ -0,0 +1,68 @@
+package grpc
+
+import (
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/datastore"
+	v1 "github.com/metal-stack/metal-api/pkg/api/v1"
+	"github.com/metal-stack/metal-lib/zapup"
+	"go.uber.org/zap"
+	"time"
+)
+
+type WaitServer struct {
+	ds     *datastore.RethinkStore
+	logger *zap.SugaredLogger
+}
+
+func NewWaitServer(ds *datastore.RethinkStore) *WaitServer {
+	return &WaitServer{
+		ds:     ds,
+		logger: zapup.MustRootLogger().Sugar(),
+	}
+}
+
+func (s *WaitServer) Wait(req *v1.WaitRequest, srv v1.Wait_WaitServer) error {
+	s.logger.Infof("wait for allocation called by", "machineID", req.Uuid)
+	ctx := srv.Context()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			id := req.Uuid
+
+			m, err := s.ds.FindMachineByID(id)
+			if err != nil {
+				return err
+			}
+
+			allocated := m.Allocation != nil
+			resp := &v1.WaitResponse{
+				Uuid:      id,
+				Allocated: allocated,
+			}
+			err = srv.Send(resp)
+			if err != nil {
+				s.logger.Errorw("send error", "error", err)
+				return err
+			}
+
+			if allocated {
+				// Close stream regularly
+				return nil
+			}
+
+			timeout := time.Now().Add(5 * time.Second)
+			ticker := time.NewTicker(500 * time.Millisecond)
```

Comment:
im unsure if this is a good idea. this ticker queries the machine table every 0.5secs (2 times per sec). ok, after 5secs there is a timeout, but the outer loop still executes this code again.
if we have 100 machines waiting, we will have ~200 db-queries/sec to find the allocated machine. i don't know if this is a problem, but the whole thing does not looks like it scales with many machines ...

Reply:
Soon we have 300machines